File: syncdb/checkpoint.py
"""Checkpoint persistence so an interrupted sync resumes where it stopped.

The checkpoint stores the last successfully committed key value. Writes are
atomic (write to a temp file, then rename) so a crash mid-write cannot leave
a corrupt checkpoint behind.
"""

import json
import logging
import os
import tempfile

log = logging.getLogger(__name__)


class Checkpoint:
    def __init__(self, path: str):
        self.path = path

    def load(self):
        """Return the last committed key, or None if starting fresh."""
        if not os.path.exists(self.path):
            return None
        with open(self.path, "r", encoding="utf-8") as fh:
            data = json.load(fh)
        last_key = data.get("last_key")
        log.info("Resuming from checkpoint: last_key=%s (%s rows done so far)",
                 last_key, data.get("rows_processed", "?"))
        return last_key

    def save(self, last_key, rows_processed: int) -> None:
        data = {"last_key": last_key, "rows_processed": rows_processed}
        directory = os.path.dirname(os.path.abspath(self.path))
        fd, tmp_path = tempfile.mkstemp(dir=directory, prefix=".ckpt-")
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as fh:
                json.dump(data, fh)
            os.replace(tmp_path, self.path)
        except BaseException:
            if os.path.exists(tmp_path):
                os.remove(tmp_path)
            raise

    def clear(self) -> None:
        if os.path.exists(self.path):
            os.remove(self.path)
            log.info("Sync finished; checkpoint file removed.")

File: src/Checkpoint.cs
using System.Text.Json;
using Microsoft.Extensions.Logging;

namespace SyncDb;

/// <summary>
/// Checkpoint persistence so an interrupted sync resumes where it stopped.
///
/// The checkpoint stores the last successfully committed key value. Writes are
/// atomic (write to a temp file, then rename) so a crash mid-write cannot leave
/// a corrupt checkpoint behind.
/// </summary>
public sealed class Checkpoint(string path, ILogger logger)
{
    /// <summary>Return the last committed key, or null if starting fresh.</summary>
    public object? Load()
    {
        if (!File.Exists(path))
        {
            return null;
        }

        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        var root = doc.RootElement;
        var lastKey = root.TryGetProperty("last_key", out var lk)
            ? JsonValueToClr(lk)
            : null;
        var rows = root.TryGetProperty("rows_processed", out var rp) ? rp.ToString() : "?";
        logger.LogInformation(
            "Resuming from checkpoint: last_key={LastKey} ({Rows} rows done so far)",
            lastKey, rows);
        return lastKey;
    }

    public void Save(object? lastKey, long rowsProcessed)
    {
        var json = JsonSerializer.Serialize(new Dictionary<string, object?>
        {
            ["last_key"] = lastKey,
            ["rows_processed"] = rowsProcessed,
        });

        var directory = Path.GetDirectoryName(Path.GetFullPath(path))!;
        var tmp = Path.Combine(directory, $".ckpt-{Path.GetRandomFileName()}");
        try
        {
            File.WriteAllText(tmp, json);
            // Atomic replace on the same volume (mirrors Python's os.replace).
            File.Move(tmp, path, overwrite: true);
        }
        catch
        {
            if (File.Exists(tmp))
            {
                File.Delete(tmp);
            }
            throw;
        }
    }

    public void Clear()
    {
        if (File.Exists(path))
        {
            File.Delete(path);
            logger.LogInformation("Sync finished; checkpoint file removed.");
        }
    }

    /// <summary>
    /// Convert a JSON scalar back to a CLR value suitable for binding as the
    /// paging key. Numbers become <see cref="decimal"/> (covers Oracle NUMBER
    /// keys); everything else stays a string.
    /// </summary>
    private static object? JsonValueToClr(JsonElement e) => e.ValueKind switch
    {
        JsonValueKind.Null => null,
        JsonValueKind.Number => e.GetDecimal(),
        JsonValueKind.String => e.GetString(),
        JsonValueKind.True => true,
        JsonValueKind.False => false,
        _ => e.ToString(),
    };
}
